### DISCUSSION TITLE
cleanFrontend

### DISCUSSION THREAD

**Discussion by JonasMorrissey** [Siouan/frontend-gradle-plugin#272] [Q&A] [answered]
Hi there! We are using this plugin and **greatly** appreciating it!

However, there is a small thing that's quite baffling to us: the cleanFrontend task. 

By itsdocumentation (https://siouan.github.io/frontend-gradle-plugin/node-corepack-npm-pnpm-yarn-tasks#cleanFrontend), "This task does nothing by default" - but in fact, due to its dependency "Depends on: [installFrontend]", cleaning the project triggers pretty much a full install of it.
Honestly, we would much rather handle cleaning of files and directories without invoking any package manager.
We would like to remove cleanFrontend from the task graph entirely - or, ideally, not even embed it into the clean lifecycle in the first place. 

Do you by chance have any advice for us, or could you maybe clarify why invoking the package manager to delete files is something we should do as well? 

All the best!

**Comment by v1nc3n4** [comment, marked as answer]
Hi @JonasMorrissey,

Thanks a lot for your message. I understand perfectly the frustration caused by task `cleanFrontend`.

#### History

The task exists from the beginning because of my personal experience, which _might_ have been relevant years ago or not at all. When developing a Javascript application, I sometimes saw a script (i.e. in the `scripts` section of the `package.json` file) dedicated to the cleaning of the project by removing artifacts built. Then, I concluded it could help to integrate such cleaning action to the Gradle `clean` task. Running a script may require the presence of some `[dev]Dependencies`, and that's why the `cleanFrontend` task depends on the `installFrontend` task.

I wondered many times if that was a good decision, before and after. As long as building the Javascript application generates artifacts in the `project.layout.buildDirectory` directory, there's no need to develop a custom JS script for cleaning. Gradle `clean` task already removes the content of this directory and may even be customized to remove additional files or directories elsewhere.

#### Immediate workarounds

That being said, disabling the `cleanFrontend` task (`enabled = false` in build script) will not prevent Gradle to execute `installNode` and `installFrontend` tasks. The only way to limit tasks executions is to allow Gradle to perform up-to-date checks with `installFrontend` task and skip its execution under some circumstances (examples provide a starting point).

#### Finally

I can remove  the `cleanFrontend` task in the next major release. I have no idea how this task is used by the community and the impacts, but I guess it is rarely used. I would provide upgrade notes for those who were relying on the `cleanScript` property.

Would that be acceptable for you?
BR

**Comment by JonasMorrissey** [comment]
Aah, right, that makes perfect sense! To be able to seamlessly integrate existing Javascript projects into Gradle, the cleanFrontend task was indeed a good solution.

What you propose would absolutely be acceptable for us, of course! If that's okay for you, that would be great! And I guess it will be quite easy for anybody who needs such a javascript internal clean task to re-add it to their build.

**Reply by bric3** [reply]
FYI in the excalidraw plugin, I had this workaround (I'm replacing it with v10.0.0)

https://github.com/bric3/excalidraw-jetbrains-plugin/blob/9f22d51429ddaee8b6c35c27aae11823be24fdcb/excalidraw-assets/build.gradle.kts#L178-L201

```kotlin
tasks {
    clean {
        // It seems that these tasks are called upon clean
        listOf(
            installNode,
            installPackageManager,
            installFrontend,
        ).forEach { task ->
            task.get().onlyIf {
                gradle.startParameter.taskNames.run {
                    none { it.endsWith("clean") }
                }
            }
        }

        dependsOn(stopYarnServer)
        delete(
            "${projectDir}/node_modules/",
            "${projectDir}/.yarn/cache/",
            "${projectDir}/.yarn/install-state.gz",
            webappFiles,
            webappExcalidrawAssets,
        )
    }
}
```
